Dockerfile

```diff
@@ -13,46 +13,179 @@
-    && apt clean
+# By default PIP install is run without cache to make image smaller
+ARG PIP_NO_CACHE_DIR
+ENV PIP_NO_CACHE_DIR=${PIP_NO_CACHE_DIR}
+RUN echo "Pip no cache dir: ${PIP_NO_CACHE_DIR}"
 
-COPY . /opt/airflow/
+ARG PIP_VERSION
+ENV PIP_VERSION=${PIP_VERSION}
+RUN echo "Pip version: ${PIP_VERSION}"
+
+RUN pip install --upgrade pip==${PIP_VERSION}
+
+# Airflow sources change frequently but dependency onfiguration won't change that often
+# We copy setup.py and other files needed to perform setup of dependencies
+# This way cache here will only be invalidated if any of the
+# version/setup configuration change but not when airflow sources change
+COPY --chown=airflow:airflow setup.py /opt/airflow/setup.py
+COPY --chown=airflow:airflow setup.cfg /opt/airflow/setup.cfg
+
+COPY --chown=airflow:airflow airflow/version.py /opt/airflow/airflow/version.py
+COPY --chown=airflow:airflow airflow/__init__.py /opt/airflow/airflow/__init__.py
+COPY --chown=airflow:airflow airflow/bin/airflow /opt/airflow/airflow/bin/airflow
+
+# Airflow Extras installed
+ARG AIRFLOW_EXTRAS="all"
+ENV AIRFLOW_EXTRAS=${AIRFLOW_EXTRAS}
+RUN echo "Installing with extras: ${AIRFLOW_EXTRAS}."
+
+# First install only dependencies but no Apache Airflow itself
+# This way regular changes in sources of Airflow will not trigger reinstallation of all dependencies
+# And this Docker layer will be reused between builds.
+RUN pip install --no-use-pep517 -e ".[${AIRFLOW_EXTRAS}]"
+
+COPY --chown=airflow:airflow airflow/www/package.json /opt/airflow/airflow/www/package.json
+COPY --chown=airflow:airflow airflow/www/package-lock.json /opt/airflow/airflow/www/package-lock.json
 
 WORKDIR /opt/airflow/airflow/www
-RUN npm install \
-    && npm run prod
+
+# Install necessary NPM dependencies (triggered by changes in package-lock.json)
+RUN gosu airflow npm ci
+
+COPY --chown=airflow:airflow airflow/www/ /opt/airflow/airflow/www/
+
+# Package NPM for production
+RUN gosu airflow npm run prod
 
 WORKDIR /opt/airflow
-RUN set -euxo pipefail \
-    && apt update \
-    && if [ -n "${BUILD_DEPS}" ]; then apt install -y $BUILD_DEPS; fi \
-    && if [ -n "${PYTHON_DEPS}" ]; then pip install --no-cache-dir ${PYTHON_DEPS}; fi \
-    && pip install --no-cache-dir --upgrade pip==19.0.1 \
-    && pip install --no-cache-dir --no-use-pep517 -e .[$AIRFLOW_DEPS] \
-    && apt purge --auto-remove -yqq $BUILD_DEPS \
-    && apt autoremove -yqq --purge \
-    && apt clean
-
-WORKDIR $AIRFLOW_HOME
-RUN mkdir -p $AIRFLOW_HOME
-COPY scripts/docker/entrypoint.sh /entrypoint.sh
-ENTRYPOINT ["/entrypoint.sh"]
+
+# Cache for this line will be automatically invalidated if any
+# of airflow sources change
+COPY --chown=airflow:airflow . /opt/airflow/
+
+# Always add-get update/upgrade here to get latest dependencies before
+# we redo pip install
+RUN apt-get update \
+    && apt-get upgrade -y --no-install-recommends \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
```

Comment:
Having this after copying source files seems out of place/not where I would have put this.

Is there a reason this is in this image rather than, say before the first `COPY` in this image, or in the airflow-apt-deps image?

Reply:
The main reason was to make sure we always use latest dependencies and update frequently. I think that originated from our original (long time ago :) ) discussion with @Fokko - he was concerned that we will be "stuck" with some older version of apt-get dependencies, rather than latest ones. 

There must be some trigger to invalidate such apt-get and we can play a bit with it. 

I also don't find it particularly good to run apt-get upgrade after adding sources. I'd rather do it after copying "setup.py" and related files - this way we would always upgrade to latest apt dependencies together with updates to pip dependencies (which makes perfect sense).

This will be less frequent on one hand but also more stable (less risk that build starts failing because some  dependency-triggered issue independent of code change). Similarly as in case of PIP dependencies that gives IMHO the best of both worlds - fairly frequent, automated update of dependencies and relative stability - the dependencies will always be "fixed" between a change in setup.py - so no more broken builds "out-of-the-blue".

Paired with regularly daily-cron-triggered fresh rebuild from the scratch, I think we might have perfect solution that detects dependency-triggered errors quickly but does not stop contributors from working and running Travis CI tests.

I think we are aligned on that @ashb and I am happy to move it there.

@Fokko  - what do you think?

Reply:
Should we move this to before the ` COPY --chown=airflow:airflow . /opt/airflow/` line then?